contracts/router/LinkedPool.sol

```diff
@@ -0,0 +1,377 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IPausable} from "./interfaces/IPausable.sol";
+import {IndexedToken, IPoolModule} from "./interfaces/IPoolModule.sol";
+import {ILinkedPool, LimitedToken} from "./interfaces/ILinkedPool.sol";
+import {IDefaultPool} from "./interfaces/IDefaultPool.sol";
+import {Action} from "./libs/Structs.sol";
+import {TokenTree} from "./tree/TokenTree.sol";
+
+import {Ownable} from "@openzeppelin/contracts-4.5.0/access/Ownable.sol";
+import {IERC20, SafeERC20} from "@openzeppelin/contracts-4.5.0/token/ERC20/utils/SafeERC20.sol";
+
+/// LinkedPool is using an internal Token Tree to aggregate a collection of pools with correlated
+/// tokens into a single wrapper, conforming to IDefaultPool interface.
+/// The internal Token Tree allows to store up to 256 tokens, which should be enough for most use cases.
+/// Note: unlike traditional Default pools, tokens in LinkedPool could be duplicated.
+/// This contract is supposed to be used in conjunction with Synapse:Bridge:
+/// - The bridged token has index == 0, and could not be duplicated in the tree.
+/// - Other tokens (correlated to bridge token) could be duplicated in the tree. Every node token in the tree
+/// is represented by a trade path from root token to node token.
+/// > This is the reason why token could be duplicated. `nUSD -> USDC` and `nUSD -> USDT -> USDC` both represent
+/// > USDC token, but via different paths from nUSD, the bridge token.
+/// In addition to the standard IDefaultPool interface, LinkedPool also implements getters to observe the internal
+/// tree, as well as the best path finder between any two tokens in the tree.
+/// Note: LinkedPool assumes that the added pool tokens have no transfer fees enabled.
+contract LinkedPool is TokenTree, Ownable, ILinkedPool {
+    using SafeERC20 for IERC20;
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address bridgeToken) TokenTree(bridgeToken) {}
+
+    // ═════════════════════════════════════════════════ EXTERNAL ══════════════════════════════════════════════════════
+
+    /// @notice Adds a pool with `N = tokensAmount` tokens to the tree by adding N-1 new nodes
+    /// as the children of the given node. Given node needs to represent a token from the pool.
+    /// @dev `poolModule` should be set to address(this) if the pool conforms to IDefaultPool interface.
+    /// Otherwise, it should be set to the address of the contract that implements the logic for pool handling.
+    /// @param nodeIndex        The index of the node to which the pool will be added
+    /// @param pool             The address of the pool
+    /// @param poolModule       The address of the pool module
+    /// @param tokensAmount     The amount of tokens in the pool
+    function addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule,
+        uint256 tokensAmount
+    ) external onlyOwner {
```

Comment:
`onlyOwner` test?

Reply:
Added in e77a732154c5ca063dfac21bf4a2bd6c94f09e89